scripts/publish.js

```diff
@@ -0,0 +1,162 @@
+    {
+      message: "Git URL:",
+      name: "git_url",
+      type: "input"
+    },
+    {
+      message: "Branch Name:",
+      name: "branch",
+      type: "input",
+      default: "master"
+    }
+  ]);
+
+  const publishSpinner = ora("Publishing Module").start();
+
+  // Publishing to module is an async operation. The following API request adds a task to an async
+  // queue. We will perform the publish operation, and then poll for the operation to complete.
+  const createResult = await apiClient.post({
+    path: "/v2/publish-module-to-catalog/",
+    body: {
+      git_url,
+      branch,
+      org_id: defaultOrganization.id,
+      is_private: true
+    }
+  });
+
+  publishSpinner.stop();
+
+  if (!createResult.ok) {
+    invalid("Unable to publish module to catalog.");
+  }
+
+  const waitingSpinner = ora(
+    "Waiting for publish to complete. This may take a few minutes.\n\n Publishing will continue if you exit here (CTRL+C)."
+  ).start();
+
+  const taskResult = await createResult.json();
+
+  let latestProcessingResultBody;
+
+  // Poll indefinitely until the task is resolved.
+  while (true) {
+    await new Promise((resolve) => {
+      setTimeout(() => {
+        resolve();
+      }, POLL_INTERVAL);
+    });
+
+    const processingResult = await apiClient.get({
+      path: "/v2/status-of-published-module/" + taskResult.task_id
+    });
+
+    if (!processingResult.ok) {
+      waitingSpinner.stop();
+
+      // TODO - ask the BE to supply a 400 when there's a regular failure?
+      if (processingResult.status === 500) {
+        const result = await processingResult.text();
+        invalid(`
+Crowdbotics CLI has encountered an error while waiting for the publish command to complete. Publishing may still finish successfully, however you may contact support with reference id: '${taskResult.task_id}'
+Error details: ${result}
+`);
+      }
+
+      invalid(
+        `Crowdbotics CLI has encountered an error while waiting for the publish command to complete. Publishing may still finish successfully, however you may contact support with reference id: '${taskResult.task_id}'`
+      );
+      break;
+    }
+
+    latestProcessingResultBody = await processingResult.json();
+
+    // There were some instances while testing where status was coming back as undefined. This may not be an issue anymore.
+    if (!latestProcessingResultBody.status) {
+      waitingSpinner.stop();
+      invalid(
+        `Crowdbotics CLI has encountered an error while waiting for the publish command to complete. Publishing may still finish successfully, however you may contact support with reference id: '${taskResult.task_id}'`
+      );
+    }
+
+    // If we've reached an end state for the polling, end the polling loop.
+    if (!IN_PROGRESS_STATUSES.includes(latestProcessingResultBody.status)) {
+      break;
+    }
+  }
+
+  waitingSpinner.stop();
+
+  if (latestProcessingResultBody.status === "FAILURE") {
```

Comment:
I don't think this case is being hit right now, because the API responds with a 500 for a task failure.

I think in an ideal world, the API should return either a 200 or 400 if the task has failed within the parameters of the task (i.e. there's an error but things are working as expected).

That would simplify the if statements a little bit.